build-logic/src/main/kotlin/dokkabuild.gradle-plugin.gradle.kts

```diff
@@ -2,34 +2,65 @@
  * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
  */
 
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import dokkabuild.utils.formattedName
+import org.jetbrains.kotlin.gradle.dsl.JvmTarget
+import org.jetbrains.kotlin.gradle.dsl.KotlinVersion
+import org.jetbrains.kotlin.gradle.tasks.KotlinJvmCompile
 
 plugins {
     id("org.gradle.kotlin.kotlin-dsl")
     id("dokkabuild.java")
-    kotlin("jvm")
     id("dokkabuild.publish-gradle-plugin")
 }
 
-// org.gradle.kotlin.kotlin-dsl sets languageVersion and apiVersion to 1.8 by default starting from Gradle 8.
-// As we need to be compatible with previous Gradle versions, we need to set it back to 1.4.
-// Note: we should do it directly on tasks and not via top-level `kotlin.compilerOptions`
-//       because `kotlin-dsl plugin` declares them on task level, and so top-level config is overridden
-tasks.withType<KotlinCompile>().configureEach {
+kotlin {
     compilerOptions {
-        languageVersion = dokkaBuild.kotlinLanguageLevel
-        apiVersion = dokkaBuild.kotlinLanguageLevel
-
-        freeCompilerArgs.addAll(
-            // need 1.4 support, otherwise there might be problems
-            // with Gradle 6.x (it's bundling Kotlin 1.4)
-            "-Xsuppress-version-warnings",
-            "-Xjsr305=strict",
-            "-Xskip-metadata-version-check",
-        )
+        // Must use Kotlin 1.4 to support Gradle 7
+        languageVersion = @Suppress("DEPRECATION") KotlinVersion.KOTLIN_1_4
+    }
+}
+
+tasks.compileKotlin {
+    compilerOptions {
+        // `kotlin-dsl` plugin overrides the versions at the task level,
+        // which takes priority over the `kotlin` project extension.
+        // So, fix it by manually setting the LV per-task.
+        languageVersion.set(kotlin.compilerOptions.languageVersion)
+        apiVersion.set(kotlin.compilerOptions.apiVersion)
     }
 }
 
 tasks.validatePlugins {
     enableStricterValidation = true
 }
+
+//region Java version target/compile config
+
+kotlin {
+    jvmToolchain(21)
+
+    compilerOptions {
+        freeCompilerArgs.add(
+            dokkaBuild.mainJavaVersion.map { "-Xjdk-release=${it.formattedName()}" }
+        )
+    }
+}
+
+tasks.withType<KotlinJvmCompile>().configureEach {
+    compilerOptions {
+        jvmTarget = dokkaBuild.mainJavaVersion.map {
+            JvmTarget.fromTarget(it.formattedName())
+        }
+    }
+}
+
+tasks.withType<JavaCompile>().configureEach {
+    options.release = dokkaBuild.mainJavaVersion.map { it.asInt() }
+}
+
+tasks.withType<Test>().configureEach {
+    javaLauncher = javaToolchains.launcherFor {
+        languageVersion = dokkaBuild.mainJavaVersion
```

Comment:
we do the same in `dokkabuild.java` but with `testJavaLauncherVersion` - is it correct?
I was thinking, that the main reason to use `testJavaLauncherVersion` is to be able to run tests on CI with different JDK versions, but now we will run them with just JDK 8.
I'm not saying this is required, just curious if it's what expected

Reply:
You're right, thanks for spotting it. Updated so that tests use `testJavaLauncherVersion`.